### PR TITLE
fix: always emit stopped on proc.close

### PR DESCRIPTION
**Pull request by ryanio**
#### What does it do?

* Emit `stopped` inside `proc.close` on all exit codes
* Adds emitting `disconnected` state if not `stopping` or `stopped`
* Adds `stopping` state
* Fixes relaying `pluginError` to UI
* Shows plugin errors in nano
* Connects `onStart` to fire on either first stdout or stderr (for example, geth=stderr and ipfs=stdout)

---

nano error

---

Fixes https://github.com/ethereum/grid/issues/355

### CODE REVIEW

File: ethereum_clients/ControlledProcess.js [hunk 1]
```diff
@@ -64,20 +64,18 @@ class ControlledProcess extends EventEmitter {
       })
 
       proc.on('close', code => {
-        if (code === 0) {
-          this.state = STATES.STOPPED
-          this.emit('newState', 'stopped')
-          this.debug('Emit newState: stopped')
-          return
+        this.state = STATES.STOPPED
```
Comment (evertonfraga) [inline ethereum_clients/ControlledProcess.js]: i am a bit skeptic that it worked well before. The behavior of arrow functions might be tricky using in combination with EventEmitter.

"It is possible to use ES6 Arrow Functions as listeners, however, when doing so, the `this` keyword will no longer reference the EventEmitter instance ([reference](https://nodejs.org/api/events.html#events_passing_arguments_and_this_to_listeners))."

here, `this` did not have the expected value.
image

**Comment by ryanio** [comment]
Good point, we are not `bind`ing it correctly as we are with other methods in the same file. I will do that now to see if it helps.

**Comment by evertonfraga** [comment]
@ryanio should the entire start process happen in a promise?

**Comment by evertonfraga** [comment]
I'm short on time, but I found some things:

1) It references a `DISCONNECTED` state that does not exist
https://github.com/ethereum/grid/blob/fixEmitStopped/ethereum_clients/ControlledProcess.js#L171
```
const STATES = {
  STARTING: 'STARTING' /* Node about to be started */,
  STARTED: 'STARTED' /* Node started */,
  CONNECTED: 'CONNECTED' /* IPC connected - all ready */,
  STOPPING: 'STOPPING' /* Node about to be stopped */,
  STOPPED: 'STOPPED' /* Node stopped */,
  ERROR: 'ERROR' /* Unexpected error */
}
```
```
this.state = STATES.DISCONNECTED
```

2) Debugging, I got some strong indications that the new state should've been emitted here as well
https://github.com/ethereum/grid/blob/fixEmitStopped/ethereum_clients/ControlledProcess.js#L142-L144

```
      this.proc.on('exit', () => {
        this.state = STATES.STOPPED
        resolve(this)
      })
```

good luck, talk soon!

**Comment by ryanio** [comment]
@evertonfraga thanks, I handled those in the latest commit

latest status: still having trouble finding why sometimes either or both grid-ui and nano don't pick up the 'stopped' event even though it is definitely emitting (or at least forwarding according to terminal)

Screen Shot 2019-07-23 at 5 49 50 PM

**Comment by evertonfraga** [comment]
Due to some unpleasant behaviors on nano with the error notifications, i am disabling those for now, which should be better handled later.

#### when stopped IPFS daemon


#### when encountered a multi-line error


#### the `geth --welp` case

**Comment by evertonfraga** [comment]
found this on grid-ui console. can debug later
image

**Comment by evertonfraga** [comment]
The error indicator on nano now pops in and out correctly and the general feeling is awesome. Good work!

**Pull request merged**